src/Microsoft.AspNetCore.Razor.Language/DefaultRazorIntermediateNodeLoweringPhase.cs

```diff
@@ -206,195 +207,220 @@ public LoweringVisitor(DocumentIntermediateNode document, IntermediateNodeBuilde
+                            Directive = descriptor,
+                            Source = BuildSourceSpanFromNode(node),
+                        };
+                    }
+                    else
                     {
-                        DirectiveName = chunkGenerator.Descriptor.Directive,
-                        Directive = chunkGenerator.Descriptor,
-                        Source = BuildSourceSpanFromNode(block),
-                    };
-                }
+                        directiveNode = new DirectiveIntermediateNode()
+                        {
+                            DirectiveName = descriptor.Directive,
+                            Directive = descriptor,
+                            Source = BuildSourceSpanFromNode(node),
+                        };
+                    }
 
-                for (var i = 0; i < chunkGenerator.Diagnostics.Count; i++)
-                {
-                    directiveNode.Diagnostics.Add(chunkGenerator.Diagnostics[i]);
-                }
+                    for (var i = 0; i < diagnostics.Length; i++)
+                    {
+                        directiveNode.Diagnostics.Add(diagnostics[i]);
+                    }
 
-                _builder.Push(directiveNode);
+                    _builder.Push(directiveNode);
+                }
 
-                VisitDefault(block);
+                Visit(node.Body);
 
-                _builder.Pop();
-            }
+                if (descriptor != null)
+                {
+                    _builder.Pop();
+                }
 
-            public override void VisitImportSpan(AddImportChunkGenerator chunkGenerator, Span span)
-            {
-                var namespaceImport = chunkGenerator.Namespace.Trim();
-                var namespaceSpan = BuildSourceSpanFromNode(span);
-                _usings.Add(new UsingReference(namespaceImport, namespaceSpan));
+                return node;
             }
 
-            public override void VisitAddTagHelperSpan(AddTagHelperChunkGenerator chunkGenerator, Span span)
+            public override SyntaxNode VisitCSharpStatementLiteral(CSharpStatementLiteralSyntax node)
             {
-                IntermediateNode directiveNode;
-                if (IsMalformed(chunkGenerator.Diagnostics))
+                var context = node.GetSpanContext();
+                if (context == null)
                 {
-                    directiveNode = new MalformedDirectiveIntermediateNode()
-                    {
-                        DirectiveName = CSharpCodeParser.AddTagHelperDirectiveDescriptor.Directive,
-                        Directive = CSharpCodeParser.AddTagHelperDirectiveDescriptor,
-                        Source = BuildSourceSpanFromNode(span),
-                    };
+                    return base.VisitCSharpStatementLiteral(node);
                 }
-                else
+                else if (context.ChunkGenerator is DirectiveTokenChunkGenerator tokenChunkGenerator)
```

Comment:
Yes. Intermediate phase is still dependent on `ChunkGenerator`s because the old code was heavily dependent on it. I plan to remove it in the future.